controllers/cloud.redhat.com/clowdjobinvocation_controller.go

```diff
@@ -0,0 +1,330 @@
+		Name: app.Spec.EnvName,
+	}, &env)
+
+	if err != nil {
+		r.Recorder.Eventf(&jinv, "Warning", "ClowdEnvMissing", "ClowdEnv [%s] is missing; Job cannot be invoked", app.Spec.EnvName)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	r.setCompletedStatus(ctx, &jinv)
+	err = r.Client.Status().Update(ctx, &jinv)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// if the Invocation is already run, don't run again
+	// May need to handle a scenario where 1 app is complete and another isn't
+	if jinv.Status.Completed {
+		r.Log.Info("Requested jobs are completed", "jobinvocation", jinv.Name, "namespace", app.Namespace)
+		for _, i := range jinv.Status.PodNames {
+			r.Log.Info(i, "ran via jobinvocation", jinv.Name, "namespace", app.Namespace)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Walk the job names to be invoked and match in the ClowdApp Spec
+	for _, jobName := range jinv.Spec.Jobs {
+		// Match the crd.Job name to the JobTemplate in ClowdApp
+		jobContent := crd.Job{}
+		err := getJobFromName(jobName, &app, &jobContent)
+		if err != nil {
+			r.Recorder.Eventf(&app, "Error", "JobNameMissing", "ClowdApp [%s] has no job named", jinv.Spec.AppName, jobName)
+			r.Log.Info("Missing Job Definition", "jobinvocation", jinv.Spec.AppName, "namespace", app.Namespace)
+			return ctrl.Result{}, err
+		}
+
+		fullJobName := fmt.Sprintf("%v-%v-%v", app.Name, jobContent.Name, jinv.Name)
+
+		if contains(jinv.Status.PodNames, fullJobName) {
+			r.Log.Info("Job has already been invoked", "jobinvocation", fullJobName, "namespace", app.Namespace)
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// We have a match that isn't duplicated and can invoke the job
+		r.Log.Info("Invoking job", "jobinvocation", jobName, "namespace", app.Namespace)
+
+		if err := r.InvokeJob(ctx, &jobContent, &app, &env, &jinv); err != nil {
+			r.Log.Info("Job Invocation Failed", "jobinvocation", jobName, "namespace", app.Namespace)
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	r.setCompletedStatus(ctx, &jinv)
+	err = r.Client.Status().Update(ctx, &jinv)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// InvokeJob is responsible for applying the Job. It also updates and reports
+// the status of that job
+func (r *ClowdJobInvocationReconciler) InvokeJob(ctx context.Context, job *crd.Job, app *crd.ClowdApp, env *crd.ClowdEnvironment, jinv *crd.ClowdJobInvocation) error {
+	nn := types.NamespacedName{
+		Name:      fmt.Sprintf("%v-%v-%v", app.Name, job.Name, jinv.Name),
+		Namespace: app.Namespace,
+	}
+	j := batchv1.Job{}
+
+	jobType := job.Type
+	switch jobType {
+
+	case "request":
+		createJobResource(app, env, nn, job, &j)
+		if err := r.Client.Create(ctx, &j); err != nil {
+			return err
+		}
+		jinv.Status.PodNames = append(jinv.Status.PodNames, j.ObjectMeta.Labels["pod"])
+		r.Log.Info("Job Invoked Successfully", "jobinvocation", job.Name, "namespace", app.Namespace)
+
+	case "deploy":
```

Comment:
I think there's a conversation to be had here about types and if "deploy" jobs are valid usecases. Is a "deploy" job really just an initContainer?

Reply:
I think you could argue that there are not a lot of use cases for deploy jobs, but they have several key differences from init containers:

- They have their own pod template
- They can run parallel with deployment pods
- They can be re-invoked via JobInvocation

So I don't think it hurts to have them, even if they aren't used very often.

Reply:
Here's my counter to it,
1) The ClowdApp is supposed to describe the app + operational maintenance jobs etc, I personally think that having a job that is invoked once and never ever again, essentially once per environment seems odd, especially if that job definition now just sits there and is never reinvoked.
2) How do we track that it is done and will never be done again? If the Job did get deleted, what would prevent a retriggering, some kind of annotation?
3) Your point about them being re-invoked makes me think, why not just make it on request and put in a job invocation when you deploy that app. There is the ability to "wait" until an app is fully deployed before running in a oneshot, and so we could just extend this functionality as a flag on the spec saying, don't run this job until the app is ready.